### ISSUE TITLE
Intermittent test failure - test_subprocess_print

### ISSUE THREAD

**Issue by takluyver** [ipython/ipykernel#98]
Seen on Jenkins on a Linux build:
https://jenkins.jupyter.org/job/jack-of-none/142/console

OK when the build was repeated.

```
======================================================================
FAIL: printing from forked mp.Process
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/lib/python3.4/site-packages/IPython/external/decorators/_decorators.py", line 222, in knownfailer
    return f(*args, **kwargs)
  File "/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/lib/python3.4/site-packages/ipykernel/tests/test_kernel.py", line 83, in test_subprocess_print
    nt.assert_equal(stdout.count("hello"), np, stdout)
nose.proxy.AssertionError: 4 != 5 : hello 0
hello 2
hello 4
hello 1

-------------------- >> begin captured logging << --------------------
root: DEBUG: Found kernel python3 in /tmp/tmpqp6_2x_7/.local/share/jupyter/kernels
root: DEBUG: Starting kernel: ['/var/lib/jenkins/shiningpanda/jobs/92f5a430/virtualenvs/d41d8cd9/bin/python3.4', '-m', 'ipykernel', '-f', '/tmp/tmpwheu19w3.json']
root: DEBUG: Connecting to: tcp://127.0.0.1:58481
root: DEBUG: connecting shell channel to tcp://127.0.0.1:33469
root: DEBUG: Connecting to: tcp://127.0.0.1:33469
root: DEBUG: connecting iopub channel to tcp://127.0.0.1:59032
root: DEBUG: Connecting to: tcp://127.0.0.1:59032
root: DEBUG: connecting stdin channel to tcp://127.0.0.1:38519
root: DEBUG: Connecting to: tcp://127.0.0.1:38519
root: DEBUG: connecting heartbeat channel to tcp://127.0.0.1:50245
--------------------- >> end captured logging << ---------------------
```

**Comment by minrk** [comment]
Yup, I'll try to track it down. I think it's a case of the forked process being allowed to tear itself down without sending queued messages. This was hard to get consistently passing with the previous implementation, and I seem to have reintroduced the same problem with the refactor.

**Comment by minrk** [comment]
Should be fixed by #99
